### ISSUE TITLE
Components don't re-render when hash location changes

### ISSUE THREAD

**Issue by mzabriskie** [remix-run/react-router#292]
When my route changes from `/user/123` to `/user/456` my handler doesn't update to reflect the new user ID.

**Comment by alansouzati** [comment]
I'm facing a similar issue:

I access a URL, for example:

```
http://localhost:900/test
```

I click in an anchor link, so the URL changes and my page scrolls to the anchor, so far so good.

```
http://localhost:900/test#my-anchor
```

Now, i press the back button in the Browser, the location changes back to `http://localhost:900/test` but the page does not scroll back to the top.

Any ideas what's going on?

I'm using react-router `0.13.4`

**Comment by alansouzati** [comment]
Never mind folks, the problem was with my CSS overriding the default browser behavior with position absolute and overflow hidden.

**Comment by knowbody** [comment]
@alansouzati yeah the back button takes you back to the place where you came from, that's the expected behaviour.

If you want to scroll back to the top you need to handle it yourself. (`window.scrollTo(0, 0)`)

**Comment by droarty** [comment]
This may just be an update of the broken link above but an excellent resource is the LifeCycle explanation in react-router's docs:  https://github.com/rackt/react-router/blob/master/docs/guides/advanced/ComponentLifecycle.md

**Comment by alexwakeman** [comment]
The fact that hash changes don't trigger a render is extremely counter-intuitive and indicative of design flaws in this router. I am now resorting to building my own router.

**Comment by taion** [comment]
Hash changes do trigger a render, BTW.

**Comment by alexwakeman** [comment]
On first pass a hash change works, but if the hash is updated from a link, there's no change. I think this router sucks.

**Comment by r2-d3** [comment]
Yes, hash change does trigger render.
I used `componentWillReceiveProps(nextProps)` to detect specific change in hash value and scroll to a certain element using **react-scroll**

You can also observe update lifecycle events on this.

**Comment by JohnnyFun** [comment]
fwiw, I ran into the same thing. It came down to me being confused that a nested component wasn't having it's componentWillReceiveProps called. That was happening because I wasn't using <Route component=... /> all the way down.

Once I wrapped components in that who needed to do routing stuff, all was well in the land of code.

Nice work on this router guys--really makes decoupling routing from UI slick. Much appreciated.
